sdks/java/managed/src/main/java/org/apache/beam/sdk/managed/ManagedSchemaTransformTranslation.java

```diff
@@ -0,0 +1,143 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.managed;
+
+import static org.apache.beam.model.pipeline.v1.RunnerApi.FunctionSpec;
+import static org.apache.beam.sdk.managed.ManagedSchemaTransformProvider.ManagedConfig;
+import static org.apache.beam.sdk.managed.ManagedSchemaTransformProvider.ManagedSchemaTransform;
+import static org.apache.beam.sdk.util.construction.PTransformTranslation.TransformPayloadTranslator;
+
+import com.google.auto.service.AutoService;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.beam.model.pipeline.v1.SchemaApi;
+import org.apache.beam.model.pipeline.v1.SchemaAwareTransforms.ManagedSchemaTransformPayload;
+import org.apache.beam.sdk.coders.RowCoder;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.runners.AppliedPTransform;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.SchemaTranslation;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.util.construction.SdkComponents;
+import org.apache.beam.sdk.util.construction.TransformPayloadTranslatorRegistrar;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.vendor.grpc.v1p60p1.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class ManagedSchemaTransformTranslation {
+  static class ManagedSchemaTransformTranslator
+      implements TransformPayloadTranslator<ManagedSchemaTransform> {
+    private final ManagedSchemaTransformProvider provider;
+    static final Schema SCHEMA =
+        Schema.builder()
+            .addStringField("transform_identifier")
+            .addNullableStringField("config")
+            .addNullableStringField("config_url")
+            .build();
+
+    public ManagedSchemaTransformTranslator() {
+      provider = new ManagedSchemaTransformProvider(null);
+    }
+
+    @Override
+    public String getUrn() {
+      return provider.identifier();
+    }
+
+    @Override
+    @SuppressWarnings("argument")
+    public @Nullable FunctionSpec translate(
+        AppliedPTransform<?, ?, ManagedSchemaTransform> application, SdkComponents components)
+        throws IOException {
+      SchemaApi.Schema expansionSchema = SchemaTranslation.schemaToProto(SCHEMA, true);
+      ManagedConfig managedConfig = application.getTransform().getManagedConfig();
+      Row configRow = toConfigRow(application.getTransform());
+      ByteArrayOutputStream os = new ByteArrayOutputStream();
+      RowCoder.of(SCHEMA).encode(configRow, os);
+
+      return FunctionSpec.newBuilder()
+          .setUrn(getUrn())
+          .setPayload(
```

Comment:
I don't think we need to set the payload at all, everything we need is already in the configRow, right?

Reply:
Not sure I understand. The idea here is to use the `Row` available in the transform payload from the job being submitted/updated to upgrade the transform via an expansion service.

Reply:
I think the idea is to store the config row and schema information in the PTransform spec, so that the runner can inspect and use it to build an upgraded transform

Reply:
Ah, probably you meant that we can get the same information from transform annotations. I think it's cleaner to have this information in the payload than annotations (and we should get rid of annotations in the future).